pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -75,26 +82,114 @@ public static PinotLLCRealtimeSegmentManager getInstance() {
+    for (int p = 0; p < nPartitions; p++) {
+      List<String> instances = new ArrayList<>(nReplicas);
+      for (int r = 0; r < nReplicas; r++) {
+        instances.add(instanceNames.get(serverId++));
+        if (serverId == instanceNames.size()) {
+          serverId = 0;
+        }
+      }
+      znRecord.setListField(Integer.toString(p), instances);
+    }
+    writeKafkaPartitionAssignemnt(realtimeTableName, znRecord);
+    setupInitialSegments(realtimeTableName, znRecord, startOffset);
   }
 
-  public void writeKafkaPartitionAssignemnt(final String realtimeTableName, ZNRecord znRecord) {
+  protected void writeKafkaPartitionAssignemnt(final String realtimeTableName, ZNRecord znRecord) {
     final String path = makeKafkaPartitionPath(realtimeTableName);
     _propertyStore.set(path, znRecord, AccessOption.PERSISTENT);
   }
+
+  protected void setupInitialSegments(String realtimeTableName, ZNRecord partitionAssignment, long startOffset) {
+    // Map of segment names to the server-instances that hold the segment.
+    final Map<String, List<String>> idealStateEntries = new HashMap<String, List<String>>(4);
+    final Map<String, List<String>> partitionMap = partitionAssignment.getListFields();
+    final int nPartitions = partitionMap.size();
+
+    // Create one segment entry in PROPERTYSTORE for each kafka partition.
+    // Any of these may already be there, so bail out clean if they are already present.
+    List<String> paths = new ArrayList<>(nPartitions);
+    List<ZNRecord> records = new ArrayList<>(nPartitions);
+    final long now = System.currentTimeMillis();
+    final int seqNum = 0; // Initial seq number for the segments
+    for (int i = 0; i < nPartitions; i++) {
+      final List instances = partitionMap.get(Integer.toString(i));
+      LLCRealtimeSegmentZKMetadata metadata = new LLCRealtimeSegmentZKMetadata();
+      String rawTableName = TableNameBuilder.extractRawTableName(realtimeTableName);
+      LLCSegmentName llcSegmentName = new LLCSegmentName(rawTableName, i, seqNum, now);
+      final String segName = llcSegmentName.getSegmentName();
+
+      metadata.setCreationTime(now);
+      metadata.setStartOffset(startOffset);
+      metadata.setNumReplicas(instances.size());
+      metadata.setTableName(realtimeTableName);
+      metadata.setSegmentName(segName);
+      metadata.setStatus(CommonConstants.Segment.Realtime.Status.IN_PROGRESS);
+
+      ZNRecord record = metadata.toZNRecord();
+      final String znodePath = PinotRealtimeSegmentManager.getSegmentsPath() + "/" + realtimeTableName + "/" + segName;
+      paths.add(znodePath);
+      records.add(record);
+      idealStateEntries.put(segName, instances);
+    }
+
+    updateHelix(realtimeTableName, idealStateEntries, paths, records);
+  }
+
+  // Marking this method protected so it is easier to test
+  protected void updateHelix(String realtimeTableName,  final Map<String, List<String>> idealStateEntries, List<String> paths, List<ZNRecord> records) {
```

Comment:
I'd rename this to `updateHelixIdealState` so that it's clearer what is being updated in Helix